File: ios/Classes/controller/MediaStreamTrackController.swift
import Flutter

/// Controller of a `MediaStreamTrack`.
class MediaStreamTrackController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of a `MediaStreamTrack`'s' proxy.
  private var track: MediaStreamTrackProxy

  /// ID of the channel created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Event channel for communicating with Flutter side.
  private var eventChannel: FlutterEventChannel

  /// Controller of the `eventChannel` management.
  private var eventController: EventController

  /// Initializes a new `MediaStreamTrackController` for the provided
  /// `MediaStreamTrackProxy`.
  init(messenger: FlutterBinaryMessenger, track: MediaStreamTrackProxy) {
    let channelName = ChannelNameGenerator.name(
      name: "MediaStreamTrack",
      id: self.channelId
    )
    self.eventChannel = FlutterEventChannel(
      name: ChannelNameGenerator.name(
        name: "MediaStreamTrackEvent",
        id: self.channelId
      ),
      binaryMessenger: messenger
    )
    self.eventController = EventController()
    self.messenger = messenger
    self.track = track
    self.channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: messenger
    )
    self.channel.setMethodCallHandler { call, result in
      self.onMethodCall(call: call, result: result)
    }
    self.eventChannel.setStreamHandler(self.eventController)
    self.track.onEnded(cb: {
      self.eventController.sendEvent(data: [
        "event": "onEnded",
      ])
    })
  }

  /// Handles all the supported Flutter method calls for the controlled
  /// `MediaStreamTrackProxy`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "setEnabled":
      let enabled = argsMap!["enabled"] as? Bool
      self.track.setEnabled(enabled: enabled!)
      result(nil)
    case "state":
      result(self.track.state().rawValue)
    case "stop":
      self.track.stop()
      result(nil)
    case "clone":
      do {
        result(
          MediaStreamTrackController(
            messenger: self.messenger,
            track: try self.track.fork()
          )
          .asFlutterResult()
        )
      } catch {
        result(error)
      }
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller into a Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    [
      "channelId": self.channelId,
      "id": self.track.id(),
      "kind": self.track.kind().rawValue,
      "deviceId": self.track.getDeviceId(),
    ]
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/controller/MediaStreamTrackController.kt
package com.instrumentisto.medea_flutter_webrtc.controller

import com.instrumentisto.medea_flutter_webrtc.proxy.MediaStreamTrackProxy
import com.instrumentisto.medea_flutter_webrtc.utils.AnyThreadSink
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.webrtc.MediaStreamTrack

/**
 * Controller of [MediaStreamTrackProxy] functional.
 *
 * @property messenger Messenger used for creating new [MethodChannel]s.
 * @property track Underlying [MediaStreamTrackProxy] to perform [MethodCall]s on.
 */
class MediaStreamTrackController(
    private val messenger: BinaryMessenger,
    val track: MediaStreamTrackProxy
) : EventChannel.StreamHandler, Controller {
  /** [CoroutineScope] for this [MediaStreamTrackController] */
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

  /** Unique ID of the [MethodChannel] of this controller. */
  private val channelId: Long = nextChannelId()

  /** ID of the underlying [MediaStreamTrack]. */
  val id: String = track.id

  /** Channel listened for the [MethodCall]s. */
  private val chan: MethodChannel =
      MethodChannel(messenger, ChannelNameGenerator.name("MediaStreamTrack", channelId))

  /** Event channel to send all the [MediaStreamTrackProxy] events into. */
  private val eventChannel: EventChannel =
      EventChannel(messenger, ChannelNameGenerator.name("MediaStreamTrackEvent", channelId))

  /** Event sink ito send all the [MediaStreamTrackProxy] events into. */
  private var eventSink: AnyThreadSink? = null

  /** [MediaStreamTrackProxy] events observer, sending all the events to the [eventSink]. */
  private var eventObserver: MediaStreamTrackProxy.Companion.EventObserver? = null

  init {
    ControllerRegistry.register(this)

    eventObserver =
        object : MediaStreamTrackProxy.Companion.EventObserver {
          override fun onEnded() {
            eventSink?.success(mapOf("event" to "onEnded"))
          }
        }

    chan.setMethodCallHandler(this)
    eventChannel.setStreamHandler(this)
    track.addEventObserver(eventObserver!!)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "setEnabled" -> {
        val enabled: Boolean = call.argument("enabled")!!
        track.setEnabled(enabled)
        result.success(null)
      }
      "state" -> {
        result.success(track.state.value)
      }
      "width" -> {
        scope.launch { result.success(track.width()) }
      }
      "height" -> {
        scope.launch { result.success(track.height()) }
      }
      "stop" -> {
        track.stop()
        result.success(null)
      }
      "clone" -> {
        result.success(MediaStreamTrackController(messenger, track.fork()).asFlutterResult())
      }
      "dispose" -> {
        disposeInternal(false)
        result.success(null)
      }
    }
  }

  override fun onListen(obj: Any?, sink: EventChannel.EventSink?) {
    if (sink != null) {
      eventSink = AnyThreadSink(sink)
    }
  }

  override fun onCancel(obj: Any?) {
    eventChannel.setStreamHandler(null)
    eventSink?.endOfStream()
    eventSink = null
  }

  override fun dispose() {
    disposeInternal(true)
  }

  /**
   * Converts this [MediaStreamTrackController] to the Flutter's method call result.
   *
   * @return [Map] generated from this controller which can be returned to the Flutter side.
   */
  fun asFlutterResult(): Map<String, Any> =
      listOf(
              Pair("channelId", channelId),
              Pair("id", track.id),
              Pair("kind", track.kind.value),
              Pair("deviceId", track.deviceId),
              Pair("facingMode", track.facingMode?.value))
          .mapNotNull { p -> p.second?.let { Pair(p.first, it) } }
          .toMap()

  /** Releases all the allocated resources. */
  private fun disposeInternal(cancel: Boolean) {
    ControllerRegistry.unregister(this)
    scope.cancel("disposed")
    chan.setMethodCallHandler(null)
    track.stop()
    track.removeEventObserver(eventObserver!!)
    eventObserver = null

    if (cancel) {
      onCancel(null)
    }
  }
}
